### PR TITLE
Add linux sampler

### LINKED ISSUE #22203: Enable taking stack snapshots of other threads: The Linux/Android edition
Follow up on https://github.com/servo/servo/issues/16740

Code in `components/background_hang_monitor/sampler.rs`(yet to be merged from https://github.com/servo/servo/pull/21673)

The suspending/resuming of a thread is (perhaps only partly) implemented, and it is mainly accessing the instruction and frame pointers in the registers, as well as perhaps the stack walking, that needs to be (re-)implemented for Linux/Android.

I think target architectures for Linux would be `x86_64`, and for Android it would be `x86` for the emulator, and `arm` for the devices. 

It appears that [`libc`](https://docs.rs/libc/0.2.43/x86_64-unknown-linux-gnu/libc/) would be the appropriate tool in both cases. 

See also the Gecko implementation at https://dxr.mozilla.org/mozilla-central/rev/b0b856065d5b7ad2996f707e6e797d0d72afd803/tools/profiler/core/platform-linux-android.cpp#85

### PR DESCRIPTION
**Pull request by gterzian**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #22203 (github issue number if applicable).

- [x] There are tests for these changes
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22355)

### CODE REVIEW

**Comment by gterzian** [comment]
@nikhilm Do you perhaps remember how you made `unwind_sys` work? Currently it's [complaining](https://travis-ci.org/servo/servo/jobs/463292682#L2048) about a non-existing `libunwind-generic` on Ubuntu. I've tried [installing `libunwind-setjmp0`](https://travis-ci.org/servo/servo/jobs/463292682#L1094), with no luck...

**Comment by nikhilm** [comment]
I have a default chromeos container, which is Debian, and I have these packages

```
ii  libunwind-dev                     1.1-4.1                           amd64        library to determine the call-chain of a program - development
ii  libunwind8                        1.1-4.1                           amd64        library to determine the call-chain of a program - runtime
ii  libunwind8-dbg                    1.1-4.1                           amd64        library to determine the call-chain of a program - runtime
ii  libunwind8-dev                    1.1-4.1                           amd64        library to determine the call-chain of a program - development
```

Perhaps you are just missing the -dev package.

**Comment by jdm** [comment]
```
running 2 tests
test test_hang_monitoring_unregister ... ok
error: process didn't exit successfully: `/home/travis/build/servo/servo/target/debug/deps/hang_monitor_tests-73c7f05bfe8ba326` (signal: 27)
The command "./mach test-unit" exited with 101.
```

**Comment by gterzian** [comment]
Given that it's `test_hang_monitoring_unregister` that fails, and signal 27 indicates it fails in relations to the `SIGPROF` signal, I guess the error happens in the `Drop` implementation of `LinuxSampler`?

Thanks for working on this one, I'm not really setup for running Linux(even on a VM of some sort) at the moment and I let this one slide once I ran into some problems on CI :)

**Comment by jdm** [comment]
It's test_hang_monitoring that has a problem, and there's output that looks like `Profiling timer expired` which I can't figure out where it's coming from.

**Comment by jdm** [comment]
I suspect that since we run tests in parallel in different threads the signals are triggering unexpected behaviour in tests. I'll try serializing them.

**Comment by jdm** [comment]
Requires https://github.com/servo/saltfs/pull/948 to be deployed before we can merge.

**Comment by jdm** [comment]
I was able to generate https://perfht.ml/2UV6uBB with the latest changes.

**Comment by jdm** [comment]
@gterzian Do you want to review my stack walking implementation (which I wrote based on https://sfackler.github.io/rstack/doc/src/unwind/lib.rs.html and https://www.nongnu.org/libunwind/man/libunwind(3).html) or should I find somebody else?

**Comment by gterzian** [comment]
I can take a look based on my initial research in the matter, it might still be a good idea to have someone else look take a final look. Maybe someone who worked on the Gecko code?

File: components/background_hang_monitor/sampler_linux.rs [hunk 1]
```diff
@@ -28,8 +182,112 @@ impl Sampler for LinuxSampler {
...
+            SHARED_STATE
+                .msg2
+                .as_ref()
+                .unwrap()
+                .wait_through_intr()
+                .expect("msg2 failed");
+        }
+
+        let context = unsafe { &mut SHARED_STATE.context.expect("valid context") };
+        let mut native_stack = NativeStack::new();
```
Comment (gterzian) [inline components/background_hang_monitor/sampler_linux.rs]: as pointed out on CI, this one appears indeed unsused...

**Review by gterzian — COMMENTED** [review]
Just one question, otherwise it looks good at least to me. I assume the use of `unwind_sys` should reduce errors like we had with the frame pointer walk. It might be good to have another pair of eyes look at it.

**Comment by jdm** [comment]
@asajeffrey Would you mind looking over these changes?

**Comment by asajeffrey** [comment]
Looking! (Slightly delayed due to unrelated gstreamer build issues).

**Comment by jdm** [comment]
CI appears to be hitting https://bugs.launchpad.net/ubuntu/+source/libunwind/+bug/1336912 which is delightful.

**Comment by asajeffrey** [comment]
Well, I'm not sure how much use my extra pair of eyes are in this case, but it LGTM. It needs run through rustfmt, and there look to be link errors on travis (https://travis-ci.com/servo/servo/jobs/187863925):
```
"-lutil" "-lutil" "-fuse-ld=gold"
  = note: /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_stream_buffer_decode'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_index_uncompressed_size'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_index_end'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_index_buffer_decode'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_stream_footer_decode'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_index_size'
          collect2: error: ld returned 1 exit status
```

**Comment by CYBAI** [comment]
In `arm*` builders:

```
error[E0583]: file not found for module `native`
  --> /home/servo/.cargo/registry/src/github.com-1ecc6299db9ec823/unwind-sys-0.1.1/src/lib.rs:12:5
   |
12 | mod native;
   |     ^^^^^^
   |
   = help: name the file either native.rs or native/mod.rs inside the directory "/home/servo/.cargo/registry/src/github.com-1ecc6299db9ec823/unwind-sys-0.1.1/src"

error: aborting due to previous error
```

In `android-mac` and `magicleap`:

<details>

```
warning: clang: warning: argument unused during compilation: '-L/Users/servo/buildbot/slave/android-mac/build/android-toolchains/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/lib/gcc/arm-linux-androideabi/4.9.x'
   Compiling brotli-decompressor v2.1.3
error: failed to run custom build command for `unwind-sys v0.1.1`
process didn't exit successfully: `/Users/servo/buildbot/slave/android-mac/build/target/android/debug/build/unwind-sys-17a4b61b40fe7910/build-script-build` (exit code: 101)
--- stderr
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Failure { command: "\"pkg-config\" \"--libs\" \"--cflags\" \"libunwind-generic\"", output: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "Package libunwind-generic was not found in the pkg-config search path.\nPerhaps you should add the directory containing `libunwind-generic.pc\'\nto the PKG_CONFIG_PATH environment variable\nNo package \'libunwind-generic\' found\n" } }', src/libcore/result.rs:997:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: std::panicking::try::do_call
             at src/libstd/panicking.rs:308
   7: <T as core::any::Any>::type_id
             at src/libcore/panicking.rs:85
   8: core::result::unwrap_failed
             at /rustc/4c27fb19ba15a2e45485e601a79914c6280196b0/src/libcore/macros.rs:16
   9: core::result::Result<T,E>::unwrap
             at /rustc/4c27fb19ba15a2e45485e601a79914c6280196b0/src/libcore/result.rs:798
  10: build_script_build::main
             at ./build.rs:11
  11: std::rt::lang_start::{{closure}}
             at /rustc/4c27fb19ba15a2e45485e601a79914c6280196b0/src/libstd/rt.rs:64
  12: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:293
  13: panic_unwind::dwarf::eh::read_encoded_pointer
             at src/libpanic_unwind/lib.rs:87
  14: <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get
             at src/libstd/panicking.rs:272
             at src/libstd/panic.rs:388
             at src/libstd/rt.rs:48
  15: std::rt::lang_start
             at /rustc/4c27fb19ba15a2e45485e601a79914c6280196b0/src/libstd/rt.rs:64
  16: build_script_build::main

warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

In `linux-*` builders:

<details>

```
  = note: /usr/bin/ld.gold: error: cannot find -llzma
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_stream_buffer_decode'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_index_uncompressed_size'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_index_end'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_index_buffer_decode'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_stream_footer_decode'
          /usr/lib/x86_64-linux-gnu/libunwind-generic.so: error: undefined reference to 'lzma_index_size'
          collect2: error: ld returned 1 exit status
```

</details>

**Pull request merged**